src/natcap/invest/urban_nature_access.py

```diff
@@ -337,11 +349,199 @@ def execute(args):
+        greenspace_budget_path, population_path, admin_unit_vector_path,
+        target_admin_unit_vector_path):
+    """Calculate average greenspace supply per admin unit.
+
+    Note:
+        The greenspace budget raster and population raster must align
+        perfectly in terms of pixel sizes and raster dimensions, and must have
+        the same projection.
+
+    Args:
+        greenspace_budget_path (string): The path to the greenspace budget
+            raster on disk.
+        population_path (string): The path to the population raster on disk.
+        admin_unit_vector_path (string): The path to the administrative units
+            vector.
+        target_admin_unit_vector_path (string): The path to where the target
+            administrative units vector will be created on disk.
+
+    Returns:
+        ``None``
+    """
+    raster_info = pygeoprocessing.get_raster_info(greenspace_budget_path)
+
+    # Reprojecting the vector here within this function allows us to produce
+    # the vector where it is used.  Otherwise, we'd be reprojecting in a
+    # separate task and then copying it to another filepath.
+    pygeoprocessing.reproject_vector(
+        admin_unit_vector_path, raster_info['projection_wkt'],
+        target_admin_unit_vector_path, driver_name='GPKG')
+
+    target_vector = gdal.OpenEx(target_admin_unit_vector_path, gdal.GA_Update)
+    target_layer = target_vector.GetLayer()
+
+    supply_sum_fieldname = 'average_greenspace_budget'
+    supply_sum_field = ogr.FieldDefn(supply_sum_fieldname, ogr.OFTReal)
+    supply_sum_field.SetWidth(24)
+    supply_sum_field.SetPrecision(11)
+    target_layer.CreateField(supply_sum_field)
+
+    greenspace_stats = pygeoprocessing.zonal_statistics(
+        (greenspace_budget_path, 1), target_admin_unit_vector_path)
+    population_stats = pygeoprocessing.zonal_statistics(
+        (population_path, 1), target_admin_unit_vector_path)
+
+    target_layer.StartTransaction()
+    for feature in target_layer:
+        feature_id = feature.GetFID()
+
+        avg_greenspace_supply_demand = (
+            greenspace_stats[feature_id]['sum'] /
+            population_stats[feature_id]['sum'])
+
+        feature.SetField(supply_sum_fieldname, avg_greenspace_supply_demand)
+        target_layer.SetFeature(feature)
+    target_layer.CommitTransaction()
+
+    target_layer = None
+    target_vector = None
+
+
+def _calculate_per_capita_greenspace_budgets(
+        greenspace_supply_raster_path, population_raster_path,
+        greenspace_demand, target_greenspace_budget_path,
+        target_supply_demand_budget_path):
+    """Calculate per-capita greenspace budgets.
+
+    This function will produce two target rasters:
+
+        * The per-capita supply-demand budget at the pixel level
+        * The greenspace supply-demand budget at the pixel level
+
+    The latter depends on the former, so running them in a single function
+    allows us to avoid reading another raster from disk.
```

Comment:
This makes sense, though I feel like a lot of InVEST models have this kind of workflow and have historically used multiple raster calculator calls. I'm not saying at all that this should do that, but I am curious if you have any insights into using more custom iterblock routines over raster calculator calls.

Reply:
Yeah, I've been flip-flopping on whether this should be implemented with `iterblocks` or with `raster_calculator`, but now that you mention it, I think these really should be `raster_calculator` calls because:

1. The pixels don't have any spatial relevance -- it's a pure pixel-stack operation that maps inputs to outputs
2. We're not modifying a raster in-place

So I'll rework this to be `raster_calculator` calls and ping you when it's ready!

Reply:
Patched in 67ec9109